src/main/java/com/mongodb/jdbc/MongoJsonSchema.java

```diff
@@ -4,22 +4,216 @@
+        MongoJsonSchema unsimplifiedSchema = toMongoJsonSchema(baseSchema);
+        return flattenNestedAnyOfs(unsimplifiedSchema);
+    }
+
+    /**
+     * Converts a deserialized jsonSchema into a MongoJsonSchema. The MongoJsonSchema instance is
+     * semantically equivalent to the base jsonSchema, but bsonType has to be a single type
+     * otherwise the types will get pushed down in the anyOf list.
+     *
+     * @param baseSchema The base json schema.
+     * @return the corresponding MongoJsonSchema.
+     */
+    private static MongoJsonSchema toMongoJsonSchema(JsonSchema baseSchema) {
+        if (null == baseSchema) {
+            return null;
+        }
+
+        MongoJsonSchema result = new MongoJsonSchema();
+        result.properties = toMongoJsonSchemaProperties(baseSchema.properties);
+        if (null != baseSchema.anyOf) {
+            result.anyOf = new HashSet<MongoJsonSchema>();
+            for (JsonSchema baseAnyOf : baseSchema.anyOf) {
+                result.anyOf.add(toSimplifiedMongoJsonSchema(baseAnyOf));
+            }
+        }
+        result.required = baseSchema.required;
+        result.items = toSimplifiedMongoJsonSchema(baseSchema.items);
+        result.additionalProperties = baseSchema.additionalProperties;
+
+        if (baseSchema.bsonType != null) {
+            Set<String> bsonTypes = polymorphicBsonTypeToStringSet(baseSchema.bsonType);
+            //  If there are many types in the set and it can not be reduced to one type after eliminating any Null
+            //  type in the list, the types will be inserted in the list of anyOf to be handled as polymorphic type.
+            if (bsonTypes.size() > 0 && bsonTypes.size() <= 2) {
+                List<String> trimmedList =
+                        bsonTypes
+                                .stream()
+                                .filter(t -> !t.equalsIgnoreCase(BSON_NULL.getBsonName()))
+                                .collect(Collectors.toList());
+
+                if (trimmedList.size() == 1) {
+                    String type = trimmedList.get(0);
+                    if (BSON_ARRAY.getBsonName().equalsIgnoreCase(type)
+                            && (null == baseSchema.items)) {
+                        // The bson type is an array of unknowns items.
+                        // The bson type is unknown and it's equivalent to any.
+                        // Return an empty schema.
+                        return new MongoJsonSchema();
+                    } else {
+                        result.bsonType = type;
+                    }
+                    return result;
+                }
+            }
+
+            // We'll need to add the types from bsontype set to the anyOf list.
+            // If the list is null, create a new one.
+            if (null == result.anyOf) {
+                result.anyOf = new HashSet<MongoJsonSchema>();
+            }
+
+            /** Push down each bsontype into its own anyOf schema.
```

Comment:
```suggestion
            /** Push down each bsontype from a bsontype array into a separate bsonType in an anyOf schema.
```

I think my problem with this comment before is "its own anyOf" which seems like we would have one `anyOf` per `bsonType`.